### ISSUE TITLE
clang does not call derived virtual class function when invoked from pointer to class member

### ISSUE THREAD

**Issue by mwittgen** [llvm/llvm-project#57593]
Code to reproduce behavior below. 
This behavior differs between `clang` and clang based compilers like `ipcx` and `gcc`/`MSVC`
The below code works for clang versions `<4.0.0` and older Intel compiler versions
Broken code: https://godbolt.org/z/r9rnv57Ye

clang requires to make `func()` pure virtual for this to work

Working code: https://godbolt.org/z/957hKM99z

```
#include <iostream>
#include <vector>
class Base {
public:
    Base() = default;
    virtual ~Base() = default;
    virtual void func() {
        std::cout << "Base\n";
    }
};

class Derived : public Base {
public:
    virtual void func() {
        std::cout << "Derived\n";
    }
};

class Test {
public:
    Test() = default;
    ~Test() = default;
    Base reg1;
    Derived reg2;
    Base Test::* r1 =   &Test::reg1;
    Base Test::* r2 = reinterpret_cast<Base Test::*>(&Test::reg2);

    void write() {
        (this->*r1).func();
        (this->*r2).func(); /* should print Derived */
        Base *b=reinterpret_cast<Base *>(&reg2);
        b->func();      
    }
};

int main() {
    Test test;
    test.write();
}
```

**Comment by dwblaikie** [comment]
I'm /guessing/ that `reinterpret_cast<Base Test::*>` is effectively UB? (probably the C++ standard says that that must point to an object of the actual type `Base` not anything derived from it - if it was valid to point to something derived, I'd guess the reinterpret cast wouldn't be required)

**Comment by dwblaikie** [comment]
Right - what I mean is, maybe that reinterpret_cast is not valid/this code isn't valid? (in the sense that maybe it causes UB)

**Comment by msebor** [comment]
The C++ standard explicitly covers this case in [§7.6.1.10 [expr.reinterpret.cast]](https://eel.is/c++draft/expr.reinterpret.cast#6):
> -5- A function pointer can be explicitly converted to a function pointer of a different type[.](https://eel.is/c++draft/expr.reinterpret.cast#6.sentence-1)
> [Note [4](https://eel.is/c++draft/expr.reinterpret.cast#note-4): The effect of calling a function through a pointer to a function type ([[dcl.fct]](https://eel.is/c++draft/dcl.fct)) that is not the same as the type used in the definition of the function is undefined ([[expr.call]](https://eel.is/c++draft/expr.call))[.](https://eel.is/c++draft/expr.reinterpret.cast#6.sentence-2) — end note]

[[conv.ptr]](https://eel.is/c++draft/conv.ptr) explains why the `reinterpret_cast` is necessary.

**Comment by Endilll** [comment]
I do not agree. Per [[basic.compound]/3](http://eel.is/c++draft/basic#compound-3.sentence-6):
> Except for pointers to static members, text referring to “pointers” does not apply to pointers to members.

Therefore, the only subclause that defines this conversion is [[expr.reinterpet.cast]/10](https://eel.is/c++draft/expr.reinterpret.cast#10). I read it as follows: "Round-trip conversions are required to yield the original value, and everything else is unspecified."

So I think we are conforming here, but we may consider changing this behavior as a quality of implementation issue.
CC @AaronBallman @shafik
